### PR TITLE
[KYUUBI #5243] Distinguish metadata between batch impl v2 and recovery

### PR DESCRIPTION
**Pull request by pan3793**
### _Why are the changes needed?_

The `recoveryMetadata` is not accurate after batch impl is introduced. This PR proposes to rename `recoveryMetadata` to `metadata` and introduce a dedicated flay `fromRecovery` to distinguish metadata between them.

This PR also partially reverts #4798, by removing unnecessary constructor parameters `shouldRunAsync` and `batchConf`

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/master/contributing/code/testing.html#running-tests) locally before make a pull request

### _Was this patch authored or co-authored using generative AI tooling?_

No.

### CODE REVIEW

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiBatchSession.scala [hunk 1]
```diff
@@ -142,32 +145,47 @@ class KyuubiBatchSession(
...
+    (metadata, fromRecovery) match {
+      case (Some(initialMetadata), false) =>
+        // new batch job created using batch impl v2
+        val metadataToUpdate = Metadata(
+          identifier = initialMetadata.identifier,
+          kyuubiInstance = connectionUrl,
+          requestName = name.orNull,
+          requestConf = optimizedConf ++ kubernetesInfo, // save the kubernetes info
+          clusterManager = batchJobSubmissionOp.builder.clusterManager())
+        sessionManager.updateMetadata(metadataToUpdate)
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiBatchSession.scala:166]: this is the key change: update metadata created by batch impl v2 before opening batch session

File: kyuubi-server/src/main/scala/org/apache/kyuubi/operation/BatchJobSubmission.scala [hunk 2]
```diff
@@ -295,19 +298,19 @@ class BatchJobSubmission(
...
+    }
+    updateBatchMetadata()
+    // TODO: add limit for max batch job submission lifetime
+    while (_applicationInfo.isDefined && !applicationTerminated(_applicationInfo)) {
+      Thread.sleep(applicationCheckInterval)
+      updateApplicationInfoMetadataIfNeeded()
+    }
+    if (applicationFailed(_applicationInfo)) {
       throw new KyuubiException(s"$batchType batch[$batchId] job failed: ${_applicationInfo}")
-    } else {
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/operation/BatchJobSubmission.scala:300]: pure refactor, no logic change here

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiBatchSession.scala [hunk 3]
```diff
@@ -43,7 +43,8 @@ class KyuubiBatchSession(
     className: String,
     batchConf: Map[String, String],
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiBatchSession.scala:44]: @turboFei Seems there is no diff between `conf` and `batchConf`, can we merge these two vars into one?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiBatchService.scala [hunk 4]
```diff
@@ -83,14 +82,9 @@ class KyuubiBatchService(
...
-              Some(metadata), // TODO some logic need to fix since it's not from recovery
+              Some(metadata),
+              false,
               shouldRunAsync = true)
-            val metadataForUpdate = Metadata(
-              identifier = batchId,
-              kyuubiInstance = kyuubiInstance,
-              requestConf = batchSession.optimizedConf,
-              clusterManager = batchSession.batchJobSubmissionOp.builder.clusterManager())
-            metadataManager.updateMetadata(metadataForUpdate, asyncRetryOnError = false)
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiBatchService.scala:93]: the logic moved to `KyuubiBatchSession#open`

File: kyuubi-server/src/main/scala/org/apache/kyuubi/operation/BatchJobSubmission.scala [hunk 5]
```diff
@@ -58,11 +58,12 @@ class BatchJobSubmission(
     className: String,
     batchConf: Map[String, String],
     batchArgs: Seq[String],
-    recoveryMetadata: Option[Metadata],
-    override val shouldRunAsync: Boolean)
+    metadata: Option[Metadata])
   extends KyuubiApplicationOperation(session) {
   import BatchJobSubmission._
 
+  override def shouldRunAsync: Boolean = true
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/operation/BatchJobSubmission.scala:65]: I wrongly thought it may be `false` on batch impl v2, in fact, it's always `true`.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiBatchSession.scala (hunk 3 above)
Comment (pan3793) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiBatchSession.scala]: merged

**Comment by pan3793** [comment]
@turboFei would you please take a look?

**Comment by pan3793** [comment]
Thanks, merged to master/1.8
